### PR TITLE
Add eslint-custom-rules package to eslint-config-fluid

### PR DESCRIPTION
**Pull request by jikim-msft**
#### Description

[ADO 6834](https://dev.azure.com/fluidframework/internal/_workitems/edit/6834/)

This PR creates a separate custom-eslint-rules module inside the `eslint-config-fluid` package. This step is necessary to use a local module as a dependency to the  `eslint-config-fluid` package which will be used across the repository. 

Pre-releasing the `eslint-config-fluid` package and testing in actual dev setting are still needed to ensure stability. 

#### Relevant Docs
- https://ronvalstar.nl/custom-local-eslint-rules
- https://pnpm.io/package_json#dependenciesmetainjected

#### `file:path` vs. `path`

Currently, every packages in FF install `@fluidframework/eslint-config-fluid` from the npm registry, whereas `tools/api-markdown-documenter` uses relative path (e.g., `"@fluidframework/eslint-config-fluid": "file:../../common/build/eslint-config-fluid"`) unlike other packages. And I believe, we are planning to use relative path for the `eslint-config-fluid` package in the future @Josmithr . 

@alexvy86 and I investigated on how the inclusion of `file:` in the `package.json` affects the linking within `node_modules`. Running `ls -l ./node_modules/@fluidframework` in `api-markdown-documenter`: 

With `file:`: `lrwxrwxrwx 1 node node 128 Jan 18 02:40 eslint-config-fluid -> ../.pnpm/file+..+..+common+build+eslint-config-fluid_bpztyfltmpuv6lhsgzfwtmxhte/node_modules/@fluidframework/eslint-config-fluid`

Without `file:`: `lrwxrwxrwx 1 node node  44 Jan 18 02:43 eslint-config-fluid -> ../../../../common/build/eslint-config-fluid`

Two things to consider about the result above:
- `pnpm i` does not complain even if `file:` is not included in the path, and rather links the package to the `node_modules` differently.  
- Relative path without `file` points to the direct location of the `devDependencies`. 

@Josmithr, I think the error that I had experienced yesterday might be due to `pnpm i` working even when  mistakenly omitting the `file:` script. Then the problem narrows down not to the instability of `pnpm` but to correctly including the `file:` script with the correct relative path.  

#### Update

`tools/api-markdown-documenter` 
Screenshot 2024-01-30 at 14 46 48

`examples/benchmarks/odspsnapshotfetch-perftestapp/src/fetchApp.ts`
Screenshot 2024-01-30 at 15 09 01

### CODE REVIEW

**Comment by Josmithr** [comment]
> @Josmithr, I think the error that I had experienced yesterday might be due to pnpm i working even when mistakenly omitting the file: script. Then the problem narrows down not to the instability of pnpm but to correctly including the file: script with the correct relative path.

Interesting. Okay. If it's working consistently with `file:`, in `eslint-config-fluid`, then I am less concerned. I'll pull down your branch and verify everything is working for me as well.

**Review by tylerbutler — COMMENTED** [review]
I'm not opposed to a separate package, but why does it need to be nested inside this package? That's strange, and strange things are hard to maintain over time. If you need a new package, create a proper package as a peer of the existing package.

**Comment by jikim-msft** [comment]
@tylerbutler 

I thought creating the `custom-rules` package inside the `eslint-config-fluid` package is preferred as the custom rules will only be used by the `eslint-config-fluid` exclusively. Due to this nature, I thought the nested structure better reflects the intention of containing any eslint-rule works to the relevant package.

In addition, I believe creating a new separate package outside the `eslint-config-fluid` was one thing that we wanted to avoid the most, thus investigating heavily on the current approach.

**Comment by tylerbutler** [comment]
Eslint plugins can export both configs and rules, and the docs indicate that the configs can include rules that are defined in the plugin. In other words, you should be need a second package.

The way we export the configs today will need to be updated:

https://eslint.org/docs/latest/extend/plugins#configs-in-plugins

Basically if we make our eslint package a proper eslint plugin then that should be a cleaner approach and address the problems of a second package.

**Comment by jikim-msft** [comment]
As per our conversation with @tylerbutler, we have decided to stick with the current approach of the PR instead of extending the configs (minimal, recommended & strict) and `fluid-custom-rules` as a plugin. 

We have concluded that creating a second package is *inevitable* due to the naming requirements of the library (Plugins must have the format of `eslint-plugin-foo` or `@scope/eslint-plugin-foo`).  For more information, read [here](https://eslint.org/docs/latest/extend/plugins#name-a-plugin).

In addition, the official ESLint documentation lacks somewhat clarity which makes it extremely difficult to devise a way of not creating a nested package.  

I have attached the pictures of `fluid-custom-rules` working expectedly if the package installs the `eslint-config-fluid` using `file:path` in the PR's description.

File: common/build/eslint-config-fluid/package.json [hunk 1]
```diff
@@ -61,5 +62,10 @@
 		"sort-json": "^2.0.1",
 		"typescript": "~5.1.6"
 	},
+	"dependenciesMeta": {
```
Comment (tylerbutler) [inline common/build/eslint-config-fluid/package.json:65]: Why is this needed?

File: common/build/eslint-config-fluid/minimal.js [hunk 2]
```diff
@@ -57,13 +57,20 @@ module.exports = {
 		"unused-imports",
 		// Plugin documentation: https://www.npmjs.com/package/eslint-plugin-unicorn
 		"unicorn",
+		// Custom ESLint rules
+		"fluid-custom-rules",
```
Comment (tylerbutler) [inline common/build/eslint-config-fluid/minimal.js:61]: I suggest using fluid-internal instead of fluid-custom-rules.

File: common/build/eslint-config-fluid/src/fluid-custom-rules/package.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+	"name": "eslint-plugin-fluid-custom-rules",
+	"version": "1.0.0",
```
Comment (Josmithr) [inline common/build/eslint-config-fluid/src/fluid-custom-rules/package.json:3]: Nit: let's make this `0.1.0` just to be safe

File: tools/api-markdown-documenter/pnpm-lock.yaml [hunk 4]
```diff
@@ -892,7 +892,7 @@ packages:
   /@rushstack/rig-package/0.5.1:
```
Comment (Josmithr) [inline tools/api-markdown-documenter/pnpm-lock.yaml:892]: We should probably revert the changes to this file, since we didn't end up making any changes in this package.

Comment (Josmithr) [reply tools/api-markdown-documenter/pnpm-lock.yaml:892]: same with the root pnpm-lock

**Review by Josmithr — COMMENTED** [review]
Looking good to me! Can you add a README.md under fluid-custom-rules explaining why we took the strange inner-package approach? It would also be nice to add some quick how-to notes for devs who want to add more custom rules in the future.

File: tools/api-markdown-documenter/pnpm-lock.yaml [hunk 5]
```diff
@@ -6535,6 +6535,12 @@ packages:
     resolution: {integrity: sha512-V50KMwwzqJV0NpZIZFwfOD5/lyny3WlSzRiXgA0G7VUnRlqttta1L6UQIHzd6EuBY/cHGfwTIck7w1yH6Q5zUw==}
     dev: true
 
+  file:../../common/build/eslint-config-fluid/src/fluid-internal:
```
Comment (jikim-msft) [inline tools/api-markdown-documenter/pnpm-lock.yaml:6538]: @Josmithr 

I think these changes with `fluid-internal` are relevant?

File: common/build/eslint-config-fluid/minimal.js (hunk 2 above)
Comment (jikim-msft) [reply common/build/eslint-config-fluid/minimal.js]: Name changed.

File: tools/api-markdown-documenter/pnpm-lock.yaml (hunk 4 above)
Comment (jikim-msft) [reply tools/api-markdown-documenter/pnpm-lock.yaml]: Lock files reverted. However, I believe the change in `api-markdown-documenter` is necessary. We could talk about this during the parking lot.

**Comment by jikim-msft** [comment]
> Looking good to me! Can you add a README.md under fluid-custom-rules explaining why we took the strange inner-package approach? It would also be nice to add some quick how-to notes for devs who want to add more custom rules in the future.

That's a great point. I have created a readme about the package explaining its purpose, background, and ways to use it. I would appreciate your feedback 😄 .

File: common/build/eslint-config-fluid/package.json (hunk 1 above)
Comment (jikim-msft) [reply common/build/eslint-config-fluid/package.json:65]: https://pnpm.io/package_json#dependenciesmetainjected

I added the field to ensure the linking between the nested packages. 

When working on it, I had also seen a suggestion from the creators of PNPM to use `dependenciesmetainjected` to ensure `pnpm i` correctly installing packages (or linking) in `node_modules`, but can't find the response at this moment. 

I can remove the field and try to verify whether everything still works or not.

File: tools/api-markdown-documenter/pnpm-lock.yaml (hunk 5 above)
Comment (Josmithr) [reply tools/api-markdown-documenter/pnpm-lock.yaml:6538]: Oh, right, because `api-markdown-documenter` has a `file:` dependency on the eslint config package. Makes sense.

File: common/build/eslint-config-fluid/src/fluid-internal/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+This package contains custom ESLint rules specific for Fluid Framework. It is a separate local module which is exported as a plugin to the `@fluidframework/eslint-config-fluid` package (consumed in `minimal.js`'s `plugins` array).
+
+An ESLint Plugin is an extension for ESLint that adds additional custom rules and configuration options.
+
+See [ESLint: Name a Plugin](https://eslint.org/docs/latest/extend/plugins) for more details.
+
+## Background
+
+The `eslint-plugin-fluid-internal` is a nested package living inside the `@fluidframework/eslint-config-fluid` package. This structure is induced by the `ESLint`'s requirements which enforce:
```
Comment (Josmithr) [inline common/build/eslint-config-fluid/src/fluid-internal/README.md:11]: Nit: we should mention that we made this choice because of this requirement **and** we don't want to have to maintain 2 separate packages.

File: common/build/eslint-config-fluid/src/fluid-internal/README.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+-   `no-restricted-tags-imports`: Restrict imports of `@internal` tagged items.
+-   `no-member-release-tags`: Restrict inclusion of any release tags inside the member of the class, function, interface, type, and enums.
+
+## Applying the Custom Rules
+
+Once creating a new custom rule, make sure to include the new rule in `minimal.js`'s `rules` field.
+
+Since `eslint-plugin-fluid-internal` is a dependency of `@fluidframework/eslint-config-fluid` package, any package using `@fluidframework/eslint-config-fluid` as its `devDependencies` will be able to apply the custom rules.
+
+Note, until the next release of the `@fluidframework/eslint-config-fluid`, in order to have custom rules working, install the local version of the package using `"@fluidframework/eslint-config-fluid": "file:relative-path"` (Check `@fluid-tools/api-markdown-documenter` as an reference).
```
Comment (Josmithr) [inline common/build/eslint-config-fluid/src/fluid-internal/README.md:30]: I would probably omit this. I don't think devs will need to know about the custom rules until we do this update and begin enforcing them across the repo.
```suggestion
```

**Review by Josmithr — APPROVED** [review]
Looks good to me, assuming everything is working correctly in testing.

Be sure to bump the major version of `eslint-config-fluid`'s package.json so we can publish a new version once this is merged.

File: common/build/eslint-config-fluid/package.json (hunk 1 above)
Comment (jikim-msft) [reply common/build/eslint-config-fluid/package.json]: It's working without the `dependenciesMeta` field so I have removed it.
